### ISSUE TITLE
--with-registry-auth flag is lost during service update

### ISSUE THREAD

**Issue by danieljuhl** [moby/moby#33929]
**Description**
I've created a docker swarm service using `--with-registry-auth`. Afterwards, when I update this service, the flag is not re-used, and has to be used explicit. Unfortunately I don't see any way to detect if a service was started with the flag.

In previous version of docker-ce (eg 17.05) it was working just fine, and the digest was fetched, and if changed the service was updated. If I explicit add `--with-registry-auth` to the second command, it works fine, and the digest is fetched.

**Steps to reproduce the issue:**
1. `docker service create --name test --with-registry-auth $(PRIVATE_REPO_WITH_AUTH_IMAGE)`
2. `docker service update --image $(PRIVATE_REPO_WITH_AUTH_IMAGE) test`

**Describe the results you received:**
The digest is removed from the image, but the service is not updated. 

**Describe the results you expected:**
I would expect the service to be updated if the digest for the image tag had changed.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Version:      17.06.0-ce
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   02c1d87
 Built:        Fri Jun 23 21:23:31 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.06.0-ce
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   02c1d87
 Built:        Fri Jun 23 21:19:04 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 5
 Running: 5
 Paused: 0
 Stopped: 0
Images: 17
Server Version: 17.06.0-ce
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 85
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: active
 NodeID: ******
 Is Manager: true
 ClusterID: ******
 Managers: 3
 Nodes: 3
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
  Force Rotate: 0
 Root Rotation In Progress: false
 Node Address: *****
 Manager Addresses:
  *****:****
  *****:****
  *****:****
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: cfb82a876ecc11b5ca0977d1733adbe58599088a
runc version: 2d41c047c83e09a6d61d464906feb2a2f3c52aa4
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-78-generic
Operating System: Ubuntu 16.04.2 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 3.859GiB
Name: node-4
ID: BAGH:2ZZ3:XGP6:NHBX:UFLG:7CEN:FPDO:NYCB:DZSX:CIZN:2QG6:OQDK
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
 provider=digitalocean
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
DigitalOcean hosted Docker swarm

**Comment by cpuguy83** [comment]
ping @nishanttotla

**Comment by nishanttotla** [comment]
Just tested and can confirm this is an issue. It seems like this is only an issue for digest fetching (the auth information does get sent to the client and the updated task still starts, albeit without a digest. @danieljuhl can you confirm this is what you see as well?). This obviously wasn't the case in 17.05 because digest fetching was being done on the daemon.

I looked at the code, and the way it works is that the CLI only sends an auth token if the `--with-registry-auth` flag was set and not if the image changed. Perhaps fixing that should be sufficient: https://github.com/docker/cli/blob/master/cli/command/service/update.go#L193

**Comment by danieljuhl** [comment]
@nishanttotla without being an experienced docker user, your description seems right. The problem is, the digest is not fetched, and thus not forwarded to the nodes running the service, and if I use a tag like "latest", the node will not update the service, as they already have a matching image. If I pass the flag along the update command, the digest is forwarded, and thus the service will be updated.

**Comment by nishanttotla** [comment]
@danieljuhl that's correct, using `latest` is a problem in this case, which is why the digest resolution exists in the first place. As far as I can tell, passing registry auth when `--image` is provided seems like a reasonable UI. We may choose to do this, but until then I recommend providing the `--with-registry-auth` flag so that your deployment works correctly.

cc @dnephin what is your opinion on the UI?

**Comment by danieljuhl** [comment]
@nishanttotla I have a few questions then:

1) Is it okay to use `--with-registry-auth` for ALL `service update` calls, even if the image is from a public registry without auth?

2) If (1) is not okay, is it then possible to know, whether a service was created with the `--with-registry-auth` flag, eg. from the service/container spec?

**Comment by dnephin** [comment]
It's fine with me, but I would check with security.

I think it should be fine to use `--with-registry-auth` for all service updates, as long as you trust the nodes on your cluster.

**Comment by nishanttotla** [comment]
@danieljuhl to answer your questions
1. Yes, it is alright to do that (for public images it will not have no effect). The only thing you might want to consider is that sending an auth token when it isn't needed might be a security concern.

2. I'm not sure about this, but I suspect the answer is no, since the auth token itself isn't displayed on inspect.

Also cc @thaJeztah @aaronlehmann should we fix this?

**Comment by thaJeztah** [comment]
I'm not sure this should check for the `sendAuth` flag to be set. If my understanding is correctly;

Passing `--with-registry-auth` on `docker service create;

- fetches the credentials from the client
- sends those credentials as part of the service create
- credentials are stored as part of the service definition

Each subsequent `docker service update` action;

- if the image is updated, or `--force-update` is used;
- looks if credentials are stored in the service's definition (/spec/whatever-it's-called)
- if so, looks up credentials needed for the image

This means that the current behaviour is that the service may use other / older credentials than are currently known by the client.

Passing `--with-registry-auth` was added because it should be an explicit action by the user to send, and store (or _update_) credentials in the service.

If we want the design to stick to that; resolving the image should use the credentials that are stored _in the service definition_, not the ones that are currently present in the client (_unless_ `--with-registry-auth` is passed). This may mean that updating the image fails if those credentials expired, or no longer valid for the requested image, in which case, it's up to the user to update the credentials with `--with-registry-auth`

**Comment by aaronlehmann** [comment]
> If we want the design to stick to that; resolving the image should use the credentials that are stored in the service definition, not the ones that are currently present in the client

These credentials are not exposed to the client. They are redacted from "service inspect" endpoint.

**Comment by thaJeztah** [comment]
@aaronlehmann wasn't digest lookup done on the daemon? Could the credentials be looked up daemon side?

**Comment by aaronlehmann** [comment]
> @aaronlehmann wasn't digest lookup done on the daemon? Could the credentials be looked up daemon side?
> 

Yes, the daemon has access to the credentials. But there was an intentional design decision not to expose them through the REST API.

The impact here is that if you change the image, you need to explicitly request that credentials for that image be used. But that seems reasonable to me. Reusing old credentials could be very problematic - for example, what happens if the new image is hosted on a different registry?

**Comment by danieljuhl** [comment]
Apologies upfront. You guys lost me a little in the very technical talk, about what actually happens under the hood.

To be very short, prior to `17.06`, updating a service image (without a digest in the command, and without the `--with-registry-auth`) which was created with `--with-registry-auth`, to new digest would be fetched and if changed, the nodes would be updated. Now, the digest is simply removed, and thus matching that image on the node, and nothing happens.

As you mention, it could be a security concern, to always append the `--with-registry-auth` flag, and thus it would make sense to have the manager handle this automatically, when needed - and in this case, have the manager fetch the digest, and update the image including the digest.

**Comment by danieljuhl** [comment]
How is the moby/docker team categorizing this issue at the moment? As a bug or feature request? 

Right now, several GUIs and automation tools for managing docker is affected. So it would be nice to know, if we can expect a change in docker, or these tools should always send --with-registry-auth in the future.

**Comment by danieljuhl** [comment]
@thaJeztah @aaronlehmann is this something you are planning to fix?

We use a small docker service which automatically updates the images when they are updated on the registry. The problem is, that even though the service/container is mounted to `/var/run/docker.sock` the `--with-registry-auth` flag doesn't work, and it gives me the following error:
```
image ${REGISTRY}/${IMAGE}:latest could not be accessed on a registry to record its digest. Each node will access ${REGISTRY}/${IMAGE}:latest independently, possibly leading to different nodes running different versions of the image.
```

All of these things worked fine in `17.05`..

**Comment by Mobe91** [comment]
@thaJeztah @aaronlehmann 
What about service updates that are not initiated by the user? E.g. one node goes down and another node goes up and the new node needs to automatically replicate a service task that went down along with the old node, i.e. it needs to pull the image from the repository. Where does the new node take the credentials from in this scenario (there is no user involvement, so no --with-registry-auth)? And what is the purpose of storing credentials in the service definition if they are likely to be outdated anyway when they are needed?
I think the best option would be to just query the credentials from the docker daemon running on the target node for a service task deployment.

I am running swarm on AWS using images from AWS ECR repositories. The repository credentials are valid for 8 hours. Each node is auto-configured with a script that updates the ECR credentials on a regular basis (see https://github.com/docker/for-aws/issues) so the local daemon always has access to valid credentials. Sadly, those credentials are not used in the scenario described above which will break rolling cluster updates (unless all credentials stored in services happen to be still valid - highly unlikely).

> This may mean that updating the image fails if those credentials expired, or no longer valid for the requested image, in which case, it's up to the user to update the credentials with --with-registry-auth

This is totally impractical for the reasons described above.
